src/looprestoration.rs

```diff
@@ -136,3 +137,160 @@ decl_looprestorationfilter_fns! {
+    if edges & LR_HAVE_BOTTOM != 0 {
+        // Copy next loop filtered rows
+        let below_1 = std::slice::from_raw_parts(lpf.offset(6 * stride as isize), unit_w as usize);
+        let below_2 = std::slice::from_raw_parts(lpf.offset(7 * stride as isize), unit_w as usize);
+
+        BD::pixel_copy(
+            &mut dst_tl[stripe_h as usize * REST_UNIT_STRIDE..],
+            below_1,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 1) as usize * REST_UNIT_STRIDE..],
+            below_2,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 2) as usize * REST_UNIT_STRIDE..],
+            below_2,
+            unit_w as usize,
+        );
+    } else {
+        // Pad with last row
+        let src = std::slice::from_raw_parts(
+            p.offset(((stripe_h - 1) as isize * stride as isize) as isize),
+            unit_w as usize,
+        );
+
+        BD::pixel_copy(
+            &mut dst_tl[stripe_h as usize * REST_UNIT_STRIDE..],
+            src,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 1) as usize * REST_UNIT_STRIDE..],
+            src,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 2) as usize * REST_UNIT_STRIDE..],
+            src,
+            unit_w as usize,
+        );
+
+        if have_left != 0 {
+            let left = &(*left.offset((stripe_h - 1) as isize))[1..];
+
+            BD::pixel_copy(&mut dst_tl[stripe_h as usize * REST_UNIT_STRIDE..], left, 3);
+            BD::pixel_copy(
+                &mut dst_tl[(stripe_h + 1) as usize * REST_UNIT_STRIDE..],
+                left,
+                3,
+            );
+            BD::pixel_copy(
+                &mut dst_tl[(stripe_h + 2) as usize * REST_UNIT_STRIDE..],
+                left,
+                3,
+            );
+        }
+    }
+
+    // Inner UNIT_WxSTRIPE_H
+    let len = (unit_w - 3 * have_left) as usize;
+    for j in 0..stripe_h as usize {
+        let p = std::slice::from_raw_parts(
+            p.offset((j * stride + 3 * have_left as usize) as isize),
+            len,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[j * REST_UNIT_STRIDE + (3 * have_left) as usize..],
+            p,
+            len,
+        );
+    }
+
+    if have_right == 0 {
+        // Pad 3x(STRIPE_H+6) with last column
+        for j in 0..stripe_h as usize + 6 {
+            let mut row_last = dst_l[(unit_w - 1) as usize + j * REST_UNIT_STRIDE];
+            let mut pad = &mut dst_l[unit_w as usize + j * REST_UNIT_STRIDE..];
+            BD::pixel_set(pad, row_last, 3);
+        }
+    }
+
+    if have_left == 0 {
+        // Pad 3x(STRIPE_H+6) with first column
+        for j in 0..stripe_h as usize + 6 {
+            let offset = j * REST_UNIT_STRIDE;
+            let val = dst[3 + offset];
```

Comment:
I think this needs a note explaining why `dst[3 + offset]` is used instead of `dst_l[offset]`.  From what I see, `dst_l` can't be used since there'd be multiple mutable borrows of `dst`, and the 3 comes from `let mut dst_l = &mut dst[(3 * (have_left == 0) as libc::c_int) as usize..];`.  Since `have_left == 0` in this branch, `3 * (have_left ==0) as libc::c_int == 3 * 1 == 3`.  Also, can you make `dst` and `dst_l` non-`mut` (the variable, not the slice)?  Then it's clear those slices don't move and the recalculation of `dst_l` is simple.